packages/edit-navigation/src/components/header/use-menu-locations.js

```diff
@@ -0,0 +1,68 @@
+/**
+ * WordPress dependencies
+ */
+import { useState, useEffect } from '@wordpress/element';
+import apiFetch from '@wordpress/api-fetch';
+import { useDispatch } from '@wordpress/data';
+
+export default function useMenuLocations() {
+	const [ menuLocationsByName, setMenuLocationsByName ] = useState( null );
+
+	useEffect( () => {
+		const fetchMenuLocationsByName = async () => {
+			const newMenuLocationsByName = await apiFetch( {
+				method: 'GET',
+				path: '/__experimental/menu-locations',
+			} );
+			setMenuLocationsByName( newMenuLocationsByName );
+		};
+		fetchMenuLocationsByName();
+	}, [] );
+
+	const { saveMenu } = useDispatch( 'core' );
+
+	const assignMenuToLocation = async ( locationName, newMenuId ) => {
+		const oldMenuId = menuLocationsByName[ locationName ].menu;
+
+		const newMenuLocationsByName = {
+			...menuLocationsByName,
+			[ locationName ]: {
+				...menuLocationsByName[ locationName ],
+				menu: newMenuId,
+			},
+		};
+
+		setMenuLocationsByName( newMenuLocationsByName );
+
+		const promises = [];
+
+		if ( oldMenuId ) {
+			promises.push(
+				saveMenu( {
+					id: oldMenuId,
+					locations: Object.values( newMenuLocationsByName )
+						.filter( ( { menu } ) => menu === oldMenuId )
+						.map( ( { name } ) => name ),
+				} )
+			);
+		}
+
+		if ( newMenuId ) {
+			promises.push(
+				saveMenu( {
+					id: newMenuId,
+					locations: Object.values( newMenuLocationsByName )
+						.filter( ( { menu } ) => menu === newMenuId )
+						.map( ( { name } ) => name ),
+				} )
+			);
+		}
+
+		await Promise.all( promises );
+	};
+
+	return [
+		menuLocationsByName ? Object.values( menuLocationsByName ) : null,
```

Comment:
`Object.values` will always return a new instance of array. Maybe memoize it?

Reply:
Does it matter? We're not passing this value as a prop or comparing it using `===`.

Reply:
My reasoning here is the same as above: https://github.com/WordPress/gutenberg/pull/25178#discussion_r486027187

If the developer ever wants to put `menuLocations` into `useEffect`, then it might cause unexpected behaviors (or even stack-overflow).